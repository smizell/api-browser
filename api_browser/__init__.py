import click
import flask.cli
import logging
import os
import webbrowser
from flask import Flask, render_template
from threading import Timer

####### Server

current_dir = os.path.dirname(__file__)
template_dir = os.path.abspath(os.path.join(current_dir, "templates"))

app = Flask(__name__, template_folder=template_dir)

# Flask shows some stuff during boot I don't want to show
# This disables that. We'll improve this later by maybe
# moving to a different HTTP app.
log = logging.getLogger("werkzeug")
log.disabled = True
flask.cli.show_server_banner = lambda *args: None


@app.route("/openapi")
def read_openapi():
    with open(app.config["OPENAPI_FILENAME"]) as f:
        content = f.read()
    return content


@app.route("/openapi-documentation")
def openapi_documentation():
    return render_template("redoc.html")


####### CLI


@click.group()
def cli():
    pass


@cli.command()
@click.argument("filename")
def openapi(filename):
    click.echo("Running API Browser server at 127.0.0.1:5000")
    click.echo("Press Ctrl+c to stop the server")
    app.config["OPENAPI_FILENAME"] = filename
    # We use a Timer so we can wait on the server to start since it blocks the thread
    Timer(0.5, _open_browser).start()
    app.run(host="127.0.0.1", port="5000", debug=False, use_reloader=False)


def _open_browser():
    webbrowser.open_new("http://127.0.0.1:5000/openapi-documentation")


if __name__ == "__main__":
    cli()
