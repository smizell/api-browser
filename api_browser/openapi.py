from typing import Optional

__all__ = ['is_ref', 'get_with_refs', 'get_schema_name']

def is_ref(value) -> bool:
    """Check if a value is a reference object (has $ref property)."""
    return isinstance(value, dict) and "$ref" in value


def get_schema_name(schema_ref: Optional[str]) -> str:
    """Extract schema name from $ref or return appropriate placeholder."""
    if not schema_ref:
        return "(none)"
    if not schema_ref.startswith("#/components/schemas/"):
        return "(inline)"
    return schema_ref.split("/")[-1]


def get_with_refs(value, path_parts, root=None, default=None):
    """
    Get a value from a nested structure, following any $ref references.
    
    Args:
        value: The value to traverse
        path_parts: List of strings/integers representing the path to follow
        root: The root document for resolving refs (defaults to value if not provided)
        default: Value to return if path is not found (defaults to None)
    
    Returns:
        The value at the path, dereferenced if it's a ref, or default if not found
    """
    if root is None:
        root = value
        
    if not path_parts:
        # If we hit a ref, resolve it and return the dereferenced value
        if is_ref(value):
            ref_path = value["$ref"].lstrip("#/").split("/")
            return get_with_refs(root, ref_path, root, default)
        return value
        
    current_key = path_parts[0]
    remaining_path = path_parts[1:]
    
    # Handle array indexing
    if isinstance(value, list):
        try:
            if not isinstance(current_key, int):
                current_key = int(current_key)
            next_value = value[current_key]
            return get_with_refs(next_value, remaining_path, root, default)
        except (ValueError, IndexError):
            return default
            
    # Handle dict access
    if isinstance(value, dict):
        # If we hit a ref, resolve it first then continue with the path
        if is_ref(value):
            ref_path = value["$ref"].lstrip("#/").split("/")
            resolved_value = get_with_refs(root, ref_path, root, default)
            return get_with_refs(resolved_value, path_parts, root, default)
            
        next_value = value.get(current_key)
        if next_value is None:
            return default
        return get_with_refs(next_value, remaining_path, root, default)
        
    return default 