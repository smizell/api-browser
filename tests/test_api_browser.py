import pytest
from api_browser import cli  # Update import to only what's needed for this test file
from click.testing import CliRunner
from api_browser import summary, schema, urls, validate_cmd  # Add urls and validate_cmd to imports
import yaml
import tempfile
import os
from pathlib import Path

# Any remaining tests for api_browser functionality would go here

def test_summary_command(snapshot):
    # Create a sample OpenAPI spec
    openapi_spec = {
        "info": {
            "title": "Test API",
            "description": "A test API description"
        },
        "paths": {
            "/pets": {
                "get": {
                    "operationId": "listPets",
                    "responses": {
                        "200": {
                            "content": {
                                "application/json": {
                                    "schema": {"$ref": "#/components/schemas/PetList"}
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    
    # Create a temporary file with the OpenAPI spec
    with tempfile.NamedTemporaryFile(mode='w', suffix='.yaml', delete=False) as f:
        yaml.dump(openapi_spec, f)
        temp_file = f.name
    
    try:
        runner = CliRunner()
        result = runner.invoke(summary, [temp_file])
        
        # Check that the command succeeded
        assert result.exit_code == 0
        
        # Compare with snapshot
        assert result.output == snapshot
        
    finally:
        # Clean up the temporary file
        os.unlink(temp_file)

def test_schema_command(snapshot):
    # Create a sample OpenAPI spec with nested schemas and circular references
    openapi_spec = {
        "paths": {
            "/users": {
                "post": {
                    "operationId": "createUser",
                    "requestBody": {
                        "content": {
                            "application/json": {
                                "schema": {"$ref": "#/components/schemas/User"}
                            }
                        }
                    },
                    "responses": {
                        "201": {
                            "content": {
                                "application/json": {
                                    "schema": {"$ref": "#/components/schemas/User"}
                                }
                            }
                        }
                    }
                },
                "get": {
                    "operationId": "listUsers",
                    "responses": {
                        "200": {
                            "content": {
                                "application/json": {
                                    "schema": {
                                        "type": "array",
                                        "items": {"$ref": "#/components/schemas/User"}
                                    }
                                }
                            }
                        }
                    }
                }
            },
            "/users/{id}": {
                "put": {
                    "operationId": "updateUser",
                    "requestBody": {
                        "content": {
                            "application/json": {
                                "schema": {"$ref": "#/components/schemas/User"}
                            }
                        }
                    },
                    "responses": {
                        "200": {
                            "content": {
                                "application/json": {
                                    "schema": {"$ref": "#/components/schemas/User"}
                                }
                            }
                        }
                    }
                }
            }
        },
        "components": {
            "schemas": {
                "Pet": {
                    "allOf": [
                        {
                            "type": "object",
                            "required": ["name"],
                            "properties": {
                                "name": {"type": "string"},
                                "age": {"type": "integer"}
                            }
                        },
                        {
                            "type": "object",
                            "properties": {
                                "owner": {"$ref": "#/components/schemas/User"}
                            }
                        }
                    ]
                },
                "Error": {
                    "oneOf": [
                        {
                            "type": "object",
                            "properties": {
                                "code": {"type": "integer"},
                                "message": {"type": "string"}
                            }
                        },
                        {
                            "type": "object",
                            "properties": {
                                "error": {"type": "string"},
                                "details": {"type": "string"}
                            }
                        }
                    ]
                },
                "Address": {
                    "type": "object",
                    "required": ["street"],
                    "properties": {
                        "street": {"type": "string"},
                        "city": {"type": "string"},
                        "country": {"type": "string"},
                        "user": {"$ref": "#/components/schemas/User"}  # Circular reference
                    }
                },
                "User": {
                    "type": "object",
                    "required": ["id", "name"],
                    "properties": {
                        "id": {"type": "integer"},
                        "name": {"type": "string"},
                        "email": {"type": "string"},
                        "address": {"$ref": "#/components/schemas/Address"},
                        "pets": {
                            "type": "array",
                            "items": {"$ref": "#/components/schemas/Pet"}
                        },
                        "metadata": {
                            "anyOf": [
                                {"type": "string"},
                                {
                                    "type": "object",
                                    "properties": {
                                        "key": {"type": "string"},
                                        "value": {"type": "string"}
                                    }
                                }
                            ]
                        }
                    }
                }
            }
        }
    }
    
    # Create a temporary file with the OpenAPI spec
    with tempfile.NamedTemporaryFile(mode='w', suffix='.yaml', delete=False) as f:
        yaml.dump(openapi_spec, f)
        temp_file = f.name
    
    try:
        runner = CliRunner()
        result = runner.invoke(schema, [temp_file, "User"])
        
        # Check that the command succeeded
        assert result.exit_code == 0
        
        # Compare with snapshot
        assert result.output == snapshot
        
    finally:
        # Clean up the temporary file
        os.unlink(temp_file)

def test_urls_command(snapshot):
    # Create a sample OpenAPI spec with various URL patterns
    openapi_spec = {
        "paths": {
            "/customers": {
                "get": {"operationId": "listCustomers"},
                "post": {"operationId": "createCustomer"}
            },
            "/customers/{id}": {
                "get": {"operationId": "getCustomer"},
                "put": {"operationId": "updateCustomer"},
                "delete": {"operationId": "deleteCustomer"}
            },
            "/customers/search": {
                "post": {"operationId": "searchCustomers"}
            },
            "/customers/{id}/orders": {
                "get": {"operationId": "listCustomerOrders"},
                "post": {"operationId": "createCustomerOrder"}
            },
            "/customers/{id}/orders/{orderId}": {
                "get": {"operationId": "getCustomerOrder"},
                "put": {"operationId": "updateCustomerOrder"},
                "delete": {"operationId": "deleteCustomerOrder"}
            },
            "/orders": {
                "get": {"operationId": "listOrders"},
                "post": {"operationId": "createOrder"}
            },
            "/orders/{id}": {
                "get": {"operationId": "getOrder"},
                "put": {"operationId": "updateOrder"},
                "delete": {"operationId": "deleteOrder"}
            }
        }
    }
    
    # Create a temporary file with the OpenAPI spec
    with tempfile.NamedTemporaryFile(mode='w', suffix='.yaml', delete=False) as f:
        yaml.dump(openapi_spec, f)
        temp_file = f.name
    
    try:
        runner = CliRunner()
        result = runner.invoke(urls, [temp_file])
        
        # Check that the command succeeded
        assert result.exit_code == 0
        
        # Compare with snapshot
        assert result.output == snapshot
        
    finally:
        # Clean up the temporary file
        os.unlink(temp_file)

def test_validate_command_valid(snapshot):
    # Create a valid OpenAPI spec
    openapi_spec = {
        "openapi": "3.0.0",
        "info": {
            "title": "Test API",
            "version": "1.0.0"
        },
        "paths": {}
    }
    
    # Create a temporary file with the OpenAPI spec
    with tempfile.NamedTemporaryFile(mode='w', suffix='.yaml', delete=False) as f:
        yaml.dump(openapi_spec, f)
        temp_file = f.name
    
    try:
        runner = CliRunner()
        result = runner.invoke(validate_cmd, [temp_file])
        
        # Check that the command succeeded
        assert result.exit_code == 0
        assert result.output.strip() == "✓ OpenAPI specification is valid"
        
        # Compare with snapshot
        assert result.output == snapshot
        
    finally:
        # Clean up the temporary file
        os.unlink(temp_file)

def test_validate_command_invalid(snapshot):
    # Create an invalid OpenAPI spec (missing required fields)
    openapi_spec = {
        "info": {
            "title": "Test API"
            # Missing required version field and openapi version
        },
        "paths": {}
    }
    
    # Create a temporary file with the OpenAPI spec
    with tempfile.NamedTemporaryFile(mode='w', suffix='.yaml', delete=False) as f:
        yaml.dump(openapi_spec, f)
        temp_file = f.name
    
    try:
        runner = CliRunner()
        result = runner.invoke(validate_cmd, [temp_file], mix_stderr=True)
        
        # Check that the command failed
        assert result.exit_code == 1
        
        # Compare with snapshot
        assert result.output == snapshot
        
    finally:
        # Clean up the temporary file
        os.unlink(temp_file)
