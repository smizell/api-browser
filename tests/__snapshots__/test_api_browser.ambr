# serializer version: 1
# name: test_schema_command
  '''
  Schema: User
  Referenced by: Address, Pet
  Requests: createUser, updateUser
  Responses: createUser, listUsers, updateUser
  
  ├── address (Address)
  │   ├── city (string)
  │   ├── country (string)
  │   ├── street* (string)
  │   └── user (User)
  │       ├── address (Address)
  │       ├── email (string)
  │       ├── id* (integer)
  │       ├── metadata (object)
  │       │   (anyOf)
  │       │       └── (string)
  │       │       └── (object)
  │       │           ├── key (string)
  │       │           └── value (string)
  │       ├── name* (string)
  │       └── pets (array[Pet])
  │           (allOf)
  │               └── (object)
  │                   ├── age (integer)
  │                   └── name* (string)
  │               └── (object)
  │                   └── owner (User)
  ├── email (string)
  ├── id* (integer)
  ├── metadata (object)
  │   (anyOf)
  │       └── (string)
  │       └── (object)
  │           ├── key (string)
  │           └── value (string)
  ├── name* (string)
  └── pets (array[Pet])
      (allOf)
          └── (object)
              ├── age (integer)
              └── name* (string)
          └── (object)
              └── owner (User)
                  ├── address (Address)
                  │   ├── city (string)
                  │   ├── country (string)
                  │   ├── street* (string)
                  │   └── user (User)
                  ├── email (string)
                  ├── id* (integer)
                  ├── metadata (object)
                  │   (anyOf)
                  │       └── (string)
                  │       └── (object)
                  │           ├── key (string)
                  │           └── value (string)
                  ├── name* (string)
                  └── pets (array[Pet])
  
  '''
# ---
# name: test_summary_command
  '''
  Title: Test API
  Description: A test API description
  
  +--------+----------+----------------+----------+------------------+-------------------+
  | Path   | Method   | Operation ID   |   Status | Request Schema   | Response Schema   |
  +========+==========+================+==========+==================+===================+
  | /pets  | GET      | listPets       |      200 | (none)           | PetList           |
  +--------+----------+----------------+----------+------------------+-------------------+
  
  '''
# ---
# name: test_urls_command
  '''
  ├── customers (createCustomer, listCustomers)
  │   ├── search (searchCustomers)
  │   └── {id} (deleteCustomer, getCustomer, updateCustomer)
  │       └── orders (createCustomerOrder, listCustomerOrders)
  │           └── {orderId} (deleteCustomerOrder, getCustomerOrder, updateCustomerOrder)
  └── orders (createOrder, listOrders)
      └── {id} (deleteOrder, getOrder, updateOrder)
  
  '''
# ---
